projects/laji-api-client-b/src/laji-api-client-b.service.ts

```diff
@@ -0,0 +1,175 @@
+import { Inject, Injectable, InjectionToken } from '@angular/core';
+import { components, operations, paths } from '../generated/api';
+import { HttpClient } from '@angular/common/http';
+import { Observable, of } from 'rxjs';
+import { shareReplay, tap } from 'rxjs/operators';
+
+type WithResponses<T> = T & { responses: unknown };
+type Parameters<T> = 'parameters' extends keyof T ? T['parameters'] : undefined;
+type ExtractContentIfExists<R> = R extends { content: { 'application/json': infer C } } ? C : null;
+type ExtractRequestBodyIfExists<R> = R extends { requestBody: { content: { 'application/json': infer C } } } ? C : never;
+type HttpSuccessCodes = 200 | 201 | 202 | 203 | 204 | 205 | 206 | 207 | 208 | 226;
+type IntersectUnionTypes<A, B> = A extends B ? A : never;
+
+type Path = keyof paths & string;
+type Method<P extends Path> = Exclude<
+  keyof { [K in keyof paths[P] as paths[P][K] extends undefined | never ? never : K]: paths[P][K] }
+, 'parameters'>;
+type Responses<P extends Path, M extends Method<P>> = WithResponses<paths[P][M]>['responses'];
+
+interface CachedValueNotStarted { _tag: 'not-started' };
+interface CachedValueLoading<T> { obs: Observable<T>; _tag: 'loading' };
+interface CachedValueCompleted<T> { val: T; lastRefresh: number; _tag: 'completed' };
+
+type CacheElement<T> = CachedValueNotStarted | CachedValueLoading<T> | CachedValueCompleted<T>;
+
+function isRecord(value: unknown): value is Record<string, unknown> {
+  return typeof value === 'object' && value !== null && !Array.isArray(value);
+}
+
+const sortRecordRecursively = (record: Record<string, unknown>): Record<string, unknown> => (
+  Object.entries(record)
+    .sort(([aKey, aVal], [bKey, bVal]) => aKey > bKey ? 1 : aKey < bKey ? -1 : 0)
+    .reduce((acc, [key, value]) => {
+      acc[key] = isRecord(value) ? sortRecordRecursively(value) : value;
+      return acc;
+    }, Object.create(null))
+);
+
+const hashRecord = (record: any): string => JSON.stringify(sortRecordRecursively(record));
+
+const splitAndResolvePath = <
+  P extends Path,
+  M extends Method<P>,
+>(path: P, params?: Parameters<paths[P][M]>): string[] => {
+  // parse path into segments based on path parameters
+  // eg. "/person/{personToken}/profile" -> ["/person", "/{personToken}", "/profile"]
+  const segments = path.split(/(\/\{[^}]+\})/).filter(Boolean);
+
+  if (!params) {
+    return segments;
+  }
+
+  // resolve path parameters (eg. replace '/{personToken}' with params.path.personToken)
+  for (let i = 1; i < segments.length; i++) {
+    const segment = segments[i];
+    segments[i] = segment.replace(/\{([^}]+)\}/, (match, p1) => (params as any)['path'][p1] ?? '');
```

Comment:
I'm a bit worried about this. An empty string path param would cause a wrong endpoint being called, which would be a disaster. Could we throw an error for undefined values and empty strings, instead of doing `?? ''`?

I think undefined values is also good to check, because we have so many cases of using non-null assertion for instance variables.

Reply:
The strict typing doesn't protect against an empty string. An empty string will cause a wrong endpoint being called (for example trying to call `GET /a/:id` but actually calls `GET /a/`, which equals `GET /a`)

And I think undefined is also still quite possible to reach here, as there are so many non-null assertions etc in the repo